ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ACSReader.scala

```diff
@@ -107,32 +119,97 @@ class FilterTableACSReader(
       .map(toIdSource)
       .pipe(mergeSort[Long])
       .statefulMapConcat(statefulDeduplicate)
-      .grouped(pageSize)
-      .map(_.toVector)
+      .via(
+        BatchN(
+          maxBatchSize = pageSize,
+          maxBatchCount = acsFetchingparallelism + 1,
+        )
+      )
       .async
+      .addAttributes(
+        Attributes.inputBuffer(initial = acsFetchingparallelism, max = acsFetchingparallelism)
+      )
       .mapAsync(acsFetchingparallelism)(fetchAcs)
   }
 }
 
 private[events] object FilterTableACSReader {
+  private val logger = ContextualizedLogger.get(this.getClass)
+
   case class Filter(party: Party, templateId: Option[Identifier])
 
+  case class IdQuery(fromExclusiveEventSeqId: Long, pageSize: Int)
+
+  case class IdQueryConfiguration(
+      minPageSize: Int,
+      maxPageSize: Int,
+  )
+
+  object IdQueryConfiguration {
+    def apply(
+        maxIdPageSize: Int,
+        idPageWorkingMemoryBytes: Int,
+        filterSize: Int,
+        idPageBufferSize: Int,
+    )(implicit loggingContext: LoggingContext): IdQueryConfiguration = {
+      val LowestIdPageSize = 10
+      // Approximation how many index entries can be present in one btree index leaf page (fetching smaller than this only adds round-trip overhead, without boiling down to smaller disk read)
+      // Experiments show party_id, template_id index has 244 tuples per page, wildcard party_id index has 254 per page (with default fill ratio for BTREE Index)
+      // Picking a lower number is for accommodating pruning, deletions, index bloat effect, which boil down to lower tuple per page ratio.
+      val RecommendedMinIdPageSize = 200
+      val calculatedMaxIdPageSize =
+        idPageWorkingMemoryBytes
+          ./(8) // IDs stored in 8 bytes
+          ./(
+            idPageBufferSize + 1
+          ) // for each filter we need one page fetched for merge sorting, and additional pages might reside in the buffer
+          ./(filterSize)
+      if (calculatedMaxIdPageSize < LowestIdPageSize) {
+        logger.warn(
+          s"Calculated maximum ID page size supporting API stream memory limits [$calculatedMaxIdPageSize] is too low: $LowestIdPageSize is used instead. Warning: API stream memory limits not respected. Warning: Dangerously low maximum ID page size can cause poor streaming performance. Filter size [$filterSize] too large?"
+        )
+        IdQueryConfiguration(LowestIdPageSize, LowestIdPageSize)
+      } else if (calculatedMaxIdPageSize < RecommendedMinIdPageSize) {
+        logger.warn(
+          s"Calculated maximum ID page size supporting API stream memory limits [$calculatedMaxIdPageSize] is very low. Warning: Low maximum ID page size can cause poor streaming performance. Filter size [$filterSize] too large?"
+        )
+        IdQueryConfiguration(calculatedMaxIdPageSize, calculatedMaxIdPageSize)
+      } else if (calculatedMaxIdPageSize < maxIdPageSize) {
+        logger.info(
```

Comment:
with current settings (working mem = 100MB, maxIdPageSize = 20000) it will be after filter size grows above, or equals 328
this means some ASX subscriptions in the NFR LR test will trigger it as well.

one could argue the maxIdPageSize is too big. Some more examples if we would lower it (maxIdPageSize, filter size after the info log is issued) for visualizing it easier:
(10000, 656) means roughly 10 contract pages per ID page
(5000, 1312) means roughly 5 contract pages per ID page
(2500, 2624) means roughly 2.5 contract pages per ID page
(1250, 5248) means roughly 1.25 contract pages per ID page

Reply:
Currently we have:
```
object IndexServiceConfig {
  val DefaultEventsPageSize: Int = 1000
  val DefaultAcsIdPageSize: Int = 20000
```
Why has the id page size been so much larger than the events page size?

I can think of these reasons:
1) limiting the number of DB queries (fewer id fetch queries),
2) making sure there are always ID's available for event fetching (fetching more ids then downstream event fetching can process at a time),
3) ids are small so we just can fit that many in memory (compared to events),
4) fetching ids is quick, so we can fetch many of them in a single query without incurring much latency.


I would also expect that fetching a single event is slower than fetching a single id, so (2) shouldn't be an inherent concern. Then it follows that decreasing the size of the id pages (~up to the size of event pages) could still keep event fetching going without starvation, while the increased number of id fetch queries would be at most the same as the number of event page queries.

Reply:
> while the increased number of id fetch queries would be at most the same as the number of event page queries.
Not exactly, this depends on the redundancy factor. Imagine 1K decomposed filters returning the same data each.

> I can think of these reasons:
1,4, and experiments shown good results with this one, but most importantly: the filter size is a multiplier on how many times we will issue this query, what was motivating to go as high as possible (turning it around: as lower we go here as higher the query load on the system).